File: leetcode/tree/257.py
# Definition for a binary tree node.
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        '''深度优先+路径记录'''
        path = []  #路径记录  全局变量，然后在dfs后pop最后一位
        ret = []

        def dfs(root: TreeNode):
            if root is None:
                return
            path.append(str(root.val))  # 前序
            if root.left is None and root.right is None:  #输出  画1～3层的树
                ret.append('->'.join(path))

            dfs(root.left)
            dfs(root.right)
            path.pop()  # 不pop的话path其实是前序遍历

        dfs(root)
        return ret

File: leetcode/tree/257.cpp
#include <iostream>
#include <vector>

#include "tree_node.h"

class Solution {
 public:
  void dfs(TreeNode* root, std::string path, std::vector<std::string>& paths) {
    if (!root) {
      return;
    }

    if (path != "") {
      path += "->";
    }
    path += std::to_string(root->val);
    if (!root->left && !root->right) {
      paths.emplace_back(path);
    }

    dfs(root->left, path, paths);
    dfs(root->right, path, paths);

    path = path.substr(0, path.size() - 2 - std::to_string(root->val).size());  // 不pop的话path其实是前序遍历
  }

  std::vector<std::string> binaryTreePaths(TreeNode* root) {
    std::vector<std::string> paths;

    dfs(root, "", paths);

    return paths;
  }
};

int main() {
  TreeNode* root = new TreeNode(1);
  root->left = new TreeNode(2);
  root->right = new TreeNode(3);
  root->left->right = new TreeNode(5);
  root->right->left = new TreeNode(6);

  Solution solution = Solution();
  auto ret = solution.binaryTreePaths(root);
  for (auto str : ret) {
    std::cout << str << std::endl;
  }
}